cmd/argo/commands/wait.go

```diff
@@ -0,0 +1,111 @@
+package commands
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	wfclient "github.com/argoproj/argo/workflow/client"
+	goversion "github.com/hashicorp/go-version"
+	"github.com/jpillora/backoff"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	RootCmd.AddCommand(waitCmd)
+	waitCmd.Flags().StringVarP(&waitArgs.namespace, "namespace", "n", "default", "Namespace in which to watch workflows")
+}
+
+type waitFlags struct {
+	namespace string
+}
+
+var waitArgs waitFlags
+
+var waitCmd = &cobra.Command{
+	Use:   "wait WORKFLOW1 WORKFLOW2..,",
+	Short: "waits for all workflows specified on command line to complete",
+	Run:   WaitWorkflows,
+}
+
+// VersionChecker checks the Kubernetes version and currently logs a message if wait should
+// be implemented using watch instead of polling.
+type VersionChecker struct{}
+
+func (vc *VersionChecker) run() {
+	// Watch APIs on CRDs using fieldSelectors are only supported in Kubernetes v1.9.0 onwards.
+	// https://github.com/kubernetes/kubernetes/issues/51046.
+	versionInfo, err := clientset.ServerVersion()
+	if err != nil {
+		log.Fatalf("Failed to get Kubernetes version: %v", err)
+	}
+
+	serverVersion, err := goversion.NewVersion(versionInfo.String())
+	if err != nil {
+		log.Fatalf("Failed to create version: %v", err)
+	}
+
+	minVersion, err := goversion.NewVersion("1.9")
+	if err != nil {
+		log.Fatalf("Failed to create minimum version: %v", err)
+	}
+
+	if serverVersion.Equal(minVersion) || serverVersion.GreaterThan(minVersion) {
+		fmt.Printf("This should be changed to use a \"watch\" based approach.\n")
+	}
+}
+
+// WorkflowStatusPoller exports methods to wait on workflows by periodically
+// querying their status.
+type WorkflowStatusPoller struct{}
+
+func (wsp *WorkflowStatusPoller) waitOnOne(wfc *wfclient.WorkflowClient, workflowName string, wg *sync.WaitGroup) bool {
+	b := &backoff.Backoff{
+		Min:    100 * time.Millisecond,
+		Max:    5 * time.Minute,
```

Comment:
I think settling on 5 minute is probably too large. The latency between when a workflow completes to when we return will be very perceivable. I think a one minute poll is more reasonable.

Similarly, the minimum of 100 milliseconds is also probably too small. Workflows will last at least a few seconds. So maybe 1 second?

Reply:
Done.